tachyon/crypto/hashes/sponge/poseidon2/poseidon2_plonky3_internal_matrix.h

```diff
@@ -7,50 +7,74 @@
 #define TACHYON_CRYPTO_HASHES_SPONGE_POSEIDON2_POSEIDON2_PLONKY3_INTERNAL_MATRIX_H_
 
 #include <numeric>
+#include <type_traits>
 
 #include "tachyon/base/logging.h"
+#include "tachyon/crypto/hashes/sponge/poseidon2/poseidon2_horizen_internal_matrix.h"
+#include "tachyon/math/finite_fields/finite_field_traits.h"
 #include "tachyon/math/matrix/matrix_types.h"
 
 namespace tachyon::crypto {
 
-template <typename PrimeField>
+template <typename F>
 class Poseidon2Plonky3InternalMatrix {
  public:
-  static void Apply(math::Vector<PrimeField>& v,
-                    const math::Vector<uint8_t>& shifts) {
+  template <typename F2 = F, std::enable_if_t<math::FiniteFieldTraits<
+                                 F2>::kIsPackedPrimeField>* = nullptr>
+  static void Apply(math::Vector<F>& v,
+                    const math::Vector<F>& diagonal_minus_one) {
+    using PrimeField =
+        std::conditional_t<math::FiniteFieldTraits<F>::kIsPackedPrimeField,
+                           typename math::FiniteFieldTraits<F>::PrimeField, F>;
+
+    Poseidon2HorizenInternalMatrix<F>::Apply(v, diagonal_minus_one);
+    if constexpr (PrimeField::Config::kUseMontgomery) {
+      static_assert(PrimeField::Config::kModulusBits <= 32);
+      for (F& f : v) {
+        f *= F::Broadcast(PrimeField::FromMontgomery(1));
+      }
+    }
+  }
+
+  template <
+      typename F2 = F,
+      std::enable_if_t<math::FiniteFieldTraits<F2>::kIsPrimeField>* = nullptr>
```

Comment:
Did you intend this function to be enabled for case where F2 is packed prime field as well?